File: iso27001-nestjs/src/api/v1/auth.controller.ts
import {
  Controller,
  Post,
  Body,
  HttpCode,
  HttpStatus,
  UnauthorizedException,
  Inject,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { IsEmail, IsString } from 'class-validator';
import { v4 as uuidv4 } from 'uuid';
import { Public } from '../../core/decorators/public.decorator';
import { USER_REPOSITORY } from '../../domain/users/user-repository.interface';
import { IUserRepository } from '../../domain/users/user-repository.interface';
import { UserService } from '../../domain/users/user.service';
import { BruteForceGuard } from '../../infrastructure/security/brute-force.guard';
import { loadConfig } from '../../core/config/configuration';
import { JwtPayload } from '../../core/auth/jwt.strategy';

/** A.14: Input validation DTOs */
class LoginDto {
  @IsEmail()
  email!: string;

  @IsString()
  password!: string;
}

class RefreshDto {
  @IsString()
  refresh_token!: string;
}

interface TokenPair {
  access_token: string;
  refresh_token: string;
  token_type: string;
}

/**
 * A.9: Authentication endpoints.
 * POST /api/v1/auth/login   — issues access + refresh token pair
 * POST /api/v1/auth/refresh — rotates token pair
 * POST /api/v1/auth/logout  — client-side only (stateless JWT)
 */
@Controller('api/v1/auth')
export class AuthController {
  constructor(
    @Inject(USER_REPOSITORY) private readonly userRepo: IUserRepository,
    private readonly userService: UserService,
    private readonly jwtService: JwtService,
    private readonly bruteForce: BruteForceGuard,
  ) {}

  @Public()
  @Post('login')
  @HttpCode(HttpStatus.OK)
  async login(@Body() dto: LoginDto): Promise<TokenPair> {
    // A.9: Brute-force check before any DB query
    await this.bruteForce.check(dto.email);

    const user = await this.userRepo.findByEmail(dto.email);
    if (!user || !(await this.userService.verifyPassword(dto.password, user.hashedPassword))) {
      await this.bruteForce.recordFailure(dto.email);
      throw new UnauthorizedException({ code: 'UNAUTHORIZED', message: 'Invalid credentials' });
    }

    if (!user.isActive) {
      throw new UnauthorizedException({ code: 'UNAUTHORIZED', message: 'User inactive' });
    }

    await this.bruteForce.clear(dto.email);
    return this.issueTokenPair(user.id, user.role);
  }

  @Post('refresh')
  @HttpCode(HttpStatus.OK)
  async refresh(@Body() dto: RefreshDto): Promise<TokenPair> {
    let payload: JwtPayload;
    try {
      payload = this.jwtService.verify<JwtPayload>(dto.refresh_token);
    } catch {
      throw new UnauthorizedException({ code: 'UNAUTHORIZED', message: 'Invalid or expired refresh token' });
    }

    if (payload.type !== 'refresh') {
      throw new UnauthorizedException({ code: 'UNAUTHORIZED', message: 'Invalid token type' });
    }

    const user = await this.userRepo.findById(payload.sub);
    if (!user || !user.isActive) {
      throw new UnauthorizedException({ code: 'UNAUTHORIZED', message: 'User not found or inactive' });
    }

    return this.issueTokenPair(user.id, user.role);
  }

  @Post('logout')
  @HttpCode(HttpStatus.OK)
  logout(): { message: string } {
    // A.9: Stateless JWT — client discards tokens.
    // For full revocation, add a Redis token denylist here.
    return { message: 'Logged out successfully' };
  }

  private issueTokenPair(userId: string, role: string): TokenPair {
    const config = loadConfig();

    const accessPayload: JwtPayload = {
      sub: userId,
      role,
      jti: uuidv4(),
      type: 'access',
    };
    const refreshPayload: JwtPayload = {
      sub: userId,
      role,
      jti: uuidv4(),
      type: 'refresh',
    };

    return {
      access_token: this.jwtService.sign(accessPayload, {
        expiresIn: config.jwtAccessExpiresIn,
      }),
      refresh_token: this.jwtService.sign(refreshPayload, {
        expiresIn: config.jwtRefreshExpiresIn,
      }),
      token_type: 'Bearer',
    };
  }
}

File: iso27001-springboot/src/main/java/com/iso27001/api/api/v1/AuthController.java
package com.iso27001.api.api.v1;

import com.iso27001.api.core.auth.JwtService;
import com.iso27001.api.domain.users.User;
import com.iso27001.api.domain.users.UserService;
import com.iso27001.api.infrastructure.security.BruteForceGuard;
import io.jsonwebtoken.Claims;
import jakarta.validation.Valid;
import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.NotBlank;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;
import java.util.Optional;

/**
 * A.9 — Authentication endpoints: login, refresh, logout.
 */
@RestController
@RequestMapping("/api/v1/auth")
public class AuthController {

    private final UserService userService;
    private final JwtService jwtService;
    private final BruteForceGuard bruteForce;

    public AuthController(UserService userService, JwtService jwtService, BruteForceGuard bruteForce) {
        this.userService = userService;
        this.jwtService = jwtService;
        this.bruteForce = bruteForce;
    }

    record LoginRequest(@Email @NotBlank String email, @NotBlank String password) {}
    record RefreshRequest(@NotBlank String refresh_token) {}
    record TokenPairResponse(String access_token, String refresh_token, String token_type) {}

    @PostMapping("/login")
    public ResponseEntity<TokenPairResponse> login(@Valid @RequestBody LoginRequest req) {
        // A.9 — brute-force check before any DB query
        bruteForce.check(req.email());

        Optional<User> userOpt = userService.findByEmail(req.email());
        if (userOpt.isEmpty() || !userService.verifyPassword(req.password(), userOpt.get().getHashedPassword())) {
            bruteForce.recordFailure(req.email());
            return ResponseEntity.status(401)
                .body(null);
        }

        User user = userOpt.get();
        if (!user.isActive()) {
            return ResponseEntity.status(401).body(null);
        }

        bruteForce.clear(req.email());
        JwtService.TokenPair pair = jwtService.issueTokenPair(user.getId().toString(), user.getRole().name());
        return ResponseEntity.ok(new TokenPairResponse(pair.accessToken(), pair.refreshToken(), "bearer"));
    }

    @PostMapping("/refresh")
    public ResponseEntity<TokenPairResponse> refresh(@Valid @RequestBody RefreshRequest req) {
        Claims claims;
        try {
            claims = jwtService.parse(req.refresh_token());
        } catch (Exception e) {
            return ResponseEntity.status(401).body(null);
        }

        if (!"refresh".equals(claims.get("type", String.class))) {
            return ResponseEntity.status(401).body(null);
        }

        Optional<User> userOpt = userService.findByEmail(claims.getSubject());
        // findById by subject (UUID)
        Optional<User> user = userOpt.isPresent() ? userOpt
            : userService.findById(java.util.UUID.fromString(claims.getSubject()));
        if (user.isEmpty() || !user.get().isActive()) {
            return ResponseEntity.status(401).body(null);
        }

        JwtService.TokenPair pair = jwtService.issueTokenPair(
            user.get().getId().toString(), user.get().getRole().name());
        return ResponseEntity.ok(new TokenPairResponse(pair.accessToken(), pair.refreshToken(), "bearer"));
    }

    @PostMapping("/logout")
    public ResponseEntity<Map<String, String>> logout() {
        // A.9 — stateless JWT: client discards tokens
        return ResponseEntity.ok(Map.of("message", "Logged out successfully"));
    }
}
